### PR TITLE
Nd convolution via blocked GEMM for C{d1,...,dn}N layout

### PR DESCRIPTION
**Pull request by csullivan**
Adds Nd cuda-c convolution reference kernel which supports padding and data/filter dilation. The kernel operates on data and filter tensors in the C{d_1,...,d_n}N and C{df_1,...,df_n}K layout for optimal coalescing and vectorization.

Kernel notes from source:
```
    // The basic strategy performed by this kernel is to convert Nd convolution
    // into a single 2D GEMM that can be block multiplied via a hierarchical strategy.
    // The spatial dimensions are squashed into a single column axis and the
    // batch number N and filter number K are the rows of A and B in the 2D GEMM
    // A * B = C, respectively. By keeping N and K in contiguous memory space,
    // coalescing and vectorization is maintained regardless of coordinate access
    // (e.g. data and filter dilation). The GEMM is broken up into shared memory tiles 
    // which are coarsened into register tiles if the batch number is a multiple 
    // of sm_tile_size * reg_tile_size ( = 8 * 4 = 32).
```
Due to the kernel complexity I added detailed comments detailing the functionality of each step.

Additional updates:
* Since nGraph tests are in N{d_1,...,d_n}C format, a transposition is required in the gpu_emitter. These should be performed via a layout pass similar to convolution in the IA transformer. This will require some updates to make the GPU transformer layout aware (c.f runtime/cpu/pass/cpu_layouts.cpp).
* op::Reshape was refactored into the CUDAEmitter.
* GPUShape::value_type was changed from uint32_t to int32_t to support the signed type `CoordinateDiff`. 

Required before merge:
* Currently register tiling is implemented when the batch and filter number are multiples of 32, but have no coverage via tests. Need to add GPU specific convolution tests for this.

Notes: 
* 4d convolution test are enabled in the GPU transformer manifest
* I would like to refactor building of the primitive arguments for convolution into a separate function, and possibly the entire op into a class. See comments in cuda_emitter.cpp.
* This kernel functions optimally for large batch sizes (multiples of 32). While this kernel extends to small batch size kernels, the Winowgrad minimal filtering technique will provide much higher performance for small batch sizes.

### CODE REVIEW

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 1]
```diff
@@ -556,15 +556,473 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
...
+                }
+                writer.block_end();
+                writer << "k++;\n";
+            }
+            writer.block_end();
+        }
+        writer.block_end();
+    }
+    writer.block_end();
+}
```
Comment (fengleitian) [inline src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: This looks good. But really complex.
One suggestion here:
Break the kernel to two: img2col for image + matrix multiplication.
The reasons are:
1. img2col is relatively simple. Even without optimization, it won't took much time.
2. matrix multiplication is already there in cuBlas and well optimized.  It take long time and need good optimization.
3. combined kernel like this are very very hard to optimize.

**Comment by csullivan** [comment]
Investigate seperation of Nd convolution into img2col and direct cuBLAS GEMM
----
Ticket: https://jira01.devtools.intel.com/browse/NGRAPH-2203

File: src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp [hunk 2]
```diff
@@ -556,15 +556,473 @@ void runtime::gpu::CudaKernelBuilder::get_replace_slice_op(
...
+           << "int* filter_str_magic, "
+           << "int* filter_str_shift"
+           << ")\n";
+    writer.block_begin();
+    {
+        writer << "Matrix* I = reinterpret_cast<Matrix*>(in);\n";
+        writer << "Matrix* F = reinterpret_cast<Matrix*>(filter);\n";
+        writer << "Matrix* O = reinterpret_cast<Matrix*>(out);\n";
+
+        writer << "__shared__ int2 lookup_table[FILTER_SIZE];\n";
```
[the remark this answers is not in the document]
Comment (csullivan) [reply src/ngraph/runtime/gpu/gpu_cuda_kernel_builder.cpp]: This is a valid concern. I did check many papers on 3d convolution and usually the filter size is not greater than 5x5x5 (e.g. V-Net https://arxiv.org/pdf/1606.04797.pdf) for training concerns. I suspect in that case the lookup table will be sufficiently small.

**Comment by csullivan** [comment]
TESTNOW

**Pull request merged**